### PR TITLE
copr: Migrate scalar function `Lpad` from TiDB 

### LINKED ISSUE #6942: UCP: Migrate scalar function `Lpad` from TiDB
## Description

Port the scalar function `Lpad` from TiDB to coprocessor.

## Score 

* 50

## Mentor(s)

* @lonng 

## Recommended Skills

* Rust programming 

## Learning Materials

Already implemented expressions ported from TiDB

- https://github.com/tikv/tikv/tree/master/components/tidb_query/src/rpn_expr) 
- https://github.com/tikv/tikv/tree/master/components/tidb_query/src/expr)

### PR DESCRIPTION
**Pull request by hsqlu**
Signed-off-by: Qiannan Lyu <lvqiannan@gmail.com>

UCP: #6942

### What problem does this PR solve?

Issue Number: close #6942

Problem Summary: Migrate scalar function `Lpad` from TiDB

### What is changed and how it works?

 Implemented scalar function Lpad

What's Changed:

### Check List 

Tests 

- Unit test

### CODE REVIEW

**Comment by hsqlu** [comment]
@lonng Can anyone please review this pr? Thanks

File: components/tidb_query_vec_expr/src/impl_string.rs [hunk 1]
```diff
@@ -116,6 +116,37 @@ pub fn rtrim(arg: &Option<Bytes>) -> Result<Option<Bytes>> {
...
+#[rpn_fn]
+#[inline]
+pub fn lpad(arg: &Option<Bytes>, len: &Option<Int>, pad: &Option<Bytes>) -> Result<Option<Bytes>> {
+    match (arg, len, pad) {
+        (Some(arg), Some(len), Some(pad)) => {
+            if *len < 0 {
+                return Ok(None);
+            }
+            let len = *len as usize;
+            if arg.len() < len && pad.is_empty() {
```
Comment (zhongzc) [inline components/tidb_query_vec_expr/src/impl_string.rs:128]: Could you add a check on whether target_len larger than the bound `MAX_BLOB_WIDTH`?

**Comment by zhongzc** [comment]
I think the normal version `lpad` which did an elegant job is worth reference.
https://github.com/tikv/tikv/blob/master/components/tidb_query_normal_expr/src/builtin_string.rs#L867

**Comment by hsqlu** [comment]
@zhongzc @lonng  PTAL, Thanks!

File: components/tikv_util/src/lib.rs [hunk 2]
```diff
@@ -49,6 +49,7 @@ pub mod logger;
 pub mod metrics;
 pub mod mpsc;
 pub mod security;
+pub mod string;
```
Comment (sticnarf) [inline components/tikv_util/src/lib.rs:52]: If it's only used in `tidb_query`, I suggest placing it in `tidb_query_shared_expr` :)

File: components/tidb_query_shared_expr/src/string.rs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+// # Examples
+// ```
+// assert_eq!(i64_to_usize(1_i64, false), (1_usize, true));
+// assert_eq!(i64_to_usize(1_i64, false), (1_usize, true));
+// assert_eq!(i64_to_usize(-1_i64, false), (1_usize, false));
+// assert_eq!(i64_to_usize(u64::max_value() as i64, true), (u64::max_value() as usize, true));
+// assert_eq!(i64_to_usize(u64::max_value() as i64, false), (1_usize, false));
+// ```
+#[inline]
+pub fn i64_to_usize(i: i64, is_unsigned: bool) -> (usize, bool) {
```
Comment (zhongzc) [inline components/tidb_query_shared_expr/src/string.rs:43]: How about moving it to `conv.rs`?

File: components/tidb_query_vec_expr/src/impl_string.rs [hunk 4]
```diff
@@ -869,6 +897,74 @@ mod tests {
         }
     }
 
+    #[test]
+    fn test_lpad() {
+        let cases = vec![
```
Comment (zhongzc) [inline components/tidb_query_vec_expr/src/impl_string.rs:902]: Can you add more corner cases, e.g. a too large `len`, a good `arg` & a good `len` but w/ an empty `pad`?

Comment (hsqlu) [reply components/tidb_query_vec_expr/src/impl_string.rs:902]: Sure

File: components/tidb_query_shared_expr/src/string.rs (hunk 3 above)
Comment (hsqlu) [reply components/tidb_query_shared_expr/src/string.rs:43]: How about the other `test_validate_target_len_for_pad` func?

Comment (zhongzc) [reply components/tidb_query_shared_expr/src/string.rs:43]: stay here is ok
